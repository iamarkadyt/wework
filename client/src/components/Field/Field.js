import React from 'react'
import DropdownList from 'react-widgets/lib/DropdownList'
import DateTimePicker from 'react-widgets/lib/DateTimePicker'
import Multiselect from 'react-widgets/lib/Multiselect'
import Moment from 'moment'
import momentLocalizer from 'react-widgets-moment'
import './Field.scss'
import { node, string, func, object, bool, number, arrayOf, oneOfType, instanceOf } from 'prop-types'

Moment.locale('en')
momentLocalizer()

const field = ({
    // different fields require different props
    type,
    name, value, onChange, // all except submit and button
    placeholder, error, // all except checkbox, submit and button
    label, // checkbox, submit, button
    list, // list, multiselect
    rows, // textarea
    disabled, // date
    onClick, style, // button, linkButton, textarea
    inline, // container
    containerStyle, // container
    children
}) => {
    let field = null
    const commonProps = { name, id: name, value, onChange }

    switch (type) {
        case "textarea":
            field = <textarea
                {...commonProps}
                placeholder={placeholder || 'Say something...'}
                className="rw-widget-container rw-widget-input"
                style={{
                    padding: '.5rem',
                    fontSize: '1rem',
                    width: '100%',
                    resize: 'none',
                    ...style
                }}
                rows={rows} />
            break
        case "date":
            field = <DateTimePicker
                {...commonProps}
                placeholder={placeholder || 'Pick a date...'}
                time={false}
                format="MMM DD, YYYY"
                editFormat="M/DD/YYYY"
                disabled={disabled} />
            break
        case "list":
            field = <DropdownList
                {...commonProps}
                placeholder={placeholder || "Choose one..."}
                data={list} />
            break
        case "multiselect":
            field = <Multiselect
                {...commonProps}
                placeholder={placeholder || "Enter a few items..."}
                data={list} />
            break
        case "checkbox":
            field = <label htmlFor={name}>
                <input type="checkbox" {...commonProps} />
                &nbsp;&nbsp;{label}
            </label>
            break
        case "submit":
            field = <button
                type="submit"
                className="button-action">
                {label || children}
            </button>
            break
        case "button":
            field = <button
                className="button"
                style={style}
                onClick={onClick}>
                {label || children}
            </button>
            break
        case "linkButton":
            field = <button
                className="link-button"
                style={style}
                onClick={onClick}>
                {label || children}
            </button>
            break
        default:
            field = <input
                {...commonProps}
                placeholder={placeholder}
                className="rw-widget-container rw-widget-input"
                type={type} />
    }

    return (
        <div style={containerStyle}
            className={['Field-container', inline ? 'inline' : ''].join(' ')}>
            {label
                && type !== "checkbox"
                && type !== "submit"
                && type !== "button"
                && type !== "linkButton"
                && <label htmlFor={name}>{label}</label>}
            {field}
            {error
                && type !== "checkbox"
                && type !== "submit"
                && type !== "button"
                && type !== "linkButton"
                && <span className="error">{error}</span>}
        </div>
    )
}

field.propTypes = {
  type: string,
  name: string,
  value: oneOfType([string, arrayOf(string), bool, instanceOf(Date)]),
  onChange: func,
  placeholder: string,
  error: string,
  label: string,
  list: arrayOf(string),
  rows: oneOfType([string, number]),
  disabled: bool,
  onClick: func,
  style: object,
  inline: bool,
  containerStyle: object,
  children: oneOfType([arrayOf(node), node])
}

export default field
