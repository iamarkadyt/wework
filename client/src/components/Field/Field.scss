@import 'variables.scss';
.Field-container {
    &.inline {
        display: inline;
    }
    width: 100%;
    margin: 1rem 0 0 0;
    input[type=checkbox] {
        transform: scale(1.5);
    }
    .button,
    button[type=submit] {
        min-width: 6rem;
        border-radius: .2rem;
        font-size: 1.1rem;
        padding: .2rem 1rem;
        border-color: $color-button-border;
        background: $color-button-backround;
        color: $color-button-font;
        &:active {
            border-color: $color-button-active-border;
            background: $color-button-active-background;
            color: $color-button-active-font;
        }
        &:hover:not(:active) {
            border-color: $color-button-hover-border;
            background: $color-button-hover-backround;
            color: $color-button-hover-font;
        }
    }
    button[type=submit] {
        float: right;
    }
    .button-action {
        background-color: darken(cornflowerblue, 10%);
        border-color: darken(cornflowerblue, 10%);
        background: linear-gradient(to bottom, darken(cornflowerblue, 10%), darken(cornflowerblue, 20%));
        color: white;
        &:active {
            background: linear-gradient(to bottom, darken(cornflowerblue, 25%), darken(cornflowerblue, 15%));
        }
        &:hover:not(:active) {
            background: linear-gradient(to bottom, darken(cornflowerblue, 8%), darken(cornflowerblue, 18%));
        }
    }
    .link-button {
        background: none;
        border: none;
        font-weight: bold;
        font-size: 1.1rem;
        color: $color-font-secondary;
        &:hover {
            color: $color-font-main;
            cursor: pointer;
        }
    }
    input[type=text],
    input[type=password] {
        font-size: 1rem;
        padding: .6rem;
        width: 100%;
    }
    label {
        display: inline-block;
        padding: .2rem;
    }
    textarea {
        background-color: $color-field-background;
        border-color: $color-field-border;
        color: $color-field-font;
    }
    .error {
        display: inline-block;
        color: $color-font-error;
        padding: .2rem;
    }
}