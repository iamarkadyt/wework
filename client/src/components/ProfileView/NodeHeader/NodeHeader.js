import React from 'react'
import Field from '../../Field/Field'
import { string, bool, func } from 'prop-types'

/**
 * Header for experience & education nodes.
 * Outputs {company} name with {timespan} or {delButton} 
 */
const NodeHeader = ({
    title,
    from,
    to,
    current,
    showDelButton,
    onDelBtnClick
}) => {
    // Month ##, ####
    const dateFormatOptions = { year: 'numeric', month: 'long', day: 'numeric' }

    from = new Date(from).toLocaleDateString('en-US', dateFormatOptions)
    to = current
        ? 'Current'
        : new Date(to).toLocaleDateString('en-US', dateFormatOptions)
    return (
        <div style={{ display: 'flex', justifyContent: 'space-between' }}>
            <h3>{title}</h3>
            {showDelButton
                ? <Field
                    type="linkButton"
                    label="Delete"
                    containerStyle={{ margin: 0, width: 'auto' }}
                    onClick={onDelBtnClick} />
                : <span>{from} â€” {to}</span>}
        </div>
    )
}

NodeHeader.propTypes = {
  title: string.isRequired,
  from: string.isRequired,
  to: string,
  current: bool,
  showDelButton: bool.isRequired,
  onDelBtnClick: func.isRequired
}

export default NodeHeader
