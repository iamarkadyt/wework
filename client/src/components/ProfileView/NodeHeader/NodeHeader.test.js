/*
 * Props
 * State
 * Interaction
 * Functions
 */

import React from 'react'
import { shallow } from 'enzyme'
import NodeHeader from './NodeHeader'
import Field from '../../Field/Field'
import cloneDeep from 'lodash.clonedeep'

let mountedComponent, props

const getMockProps = () => {
  return cloneDeep({
    title: "Some title",
    from: "Fri Dec 28 2018 18:30:47 GMT-0800 (Pacific Standard Time)",
    current: true,
    showDelButton: true,
    onDelBtnClick: jest.fn()
  })
}

// dateFormatOptions 
const dfo = { year: 'numeric', month: 'long', day: 'numeric' }

const comp = () => {
  if (!mountedComponent) {
    mountedComponent = shallow(<NodeHeader {...props} />)
  }
  return mountedComponent
}

describe('NodeHeader', () => {
  beforeEach(() => {
    mountedComponent = undefined
    props = getMockProps()
  })

  describe('props effects', () => {
    describe('[current]', () => {
      it('if true, comp renders from_date -> "Current" on date-line', () => {
        props.current = true
        props.showDelButton = false // jic
        expect(comp().find('h3+span').text()).toContain('Current')
      })

      it('if false, comp renders from_date -> to_date on date-line', () => {
        props.current = false
        props.showDelButton = false // jic
        expect(comp().find('h3+span').text()).not.toContain('Current')
      })
    })

    describe('[showDelButton]', () => {
      it('if true, comp renders delete button instead of date-line', () => {
        props.showDelButton = true
        expect(comp().find(Field).exists()).toBe(true)
      })

      it('if false, comp renders date-line', () => {
        props.showDelButton = false
        expect(comp().find(Field).exists()).toBe(false)
      })
    })
  })

  describe('props passing', () => {
    it('h3 gets props.title', () => {
      const title = 'custom title'
      props.title = title
      expect(comp().find('h3').text()).toBe(title)
    })

    describe('delete button', () => {
      it('receives correct set of props', () => {
        const expectedProps = [
          "type",
          "label",
          "containerStyle",
          "onClick"
        ]
        props.showDelButton = true
        expect(Object.keys(comp().find(Field).props())).toStrictEqual(expectedProps)
      })

      it('receives props.onDelBtnClick for onClick prop', () => {
        const fn = jest.fn()
        props.onDelBtnClick = fn
        props.showDelButton = true
        expect(comp().find(Field).prop('onClick')).toBe(fn)
      })
    })

    describe('date-line', () => {
      it('receives props.from', () => {
        const date = "Fri Dec 28 2018 18:30:47 GMT-0800 (Pacific Standard Time)"
        props.from = date
        props.showDelButton = false
        expect(comp().find('h3+span').text()).toContain(new Date(date).toLocaleDateString('en-US', dfo))
      })

      it('receives props.to if current is false', () => {
        const date = "Thu Jan 28 2030 19:00:00 GMT-0800 (Pacific Standard Time)"
        props.current = false
        props.to = date
        props.showDelButton = false
        expect(comp().find('h3+span').text()).toContain(new Date(date).toLocaleDateString('en-US', dfo))
      })
    })
  })

  describe('passes a snapshot test', () => {
    it('showing delete button', () => {
      props.showDelButton = true
      expect(comp()).toMatchSnapshot()
    })

    it('showing date-line', () => {
      props.showDelButton = false
      props.current = false
      expect(comp()).toMatchSnapshot()
    })

    it('rendering "Current" instead of to_date', () => {
      props.showDelButton = false
      props.current = true
      expect(comp()).toMatchSnapshot()
    })
  })
})
