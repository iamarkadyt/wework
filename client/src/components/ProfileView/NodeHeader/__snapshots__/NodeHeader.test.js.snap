// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeHeader passes a snapshot test rendering "Current" instead of to_date 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "justifyContent": "space-between",
    }
  }
>
  <h3>
    Some title
  </h3>
  <span>
    December 28, 2018
     — 
    Current
  </span>
</div>
`;

exports[`NodeHeader passes a snapshot test showing date-line 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "justifyContent": "space-between",
    }
  }
>
  <h3>
    Some title
  </h3>
  <span>
    December 28, 2018
     — 
    Invalid Date
  </span>
</div>
`;

exports[`NodeHeader passes a snapshot test showing delete button 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "justifyContent": "space-between",
    }
  }
>
  <h3>
    Some title
  </h3>
  <field
    containerStyle={
      Object {
        "margin": 0,
        "width": "auto",
      }
    }
    label="Delete"
    onClick={[MockFunction]}
    type="linkButton"
  />
</div>
`;
